resources/assets/lib/beatmap-discussions/editor-toolbar.tsx

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import { ReactNode, useEffect, useRef } from 'react';
+import * as React from 'react';
+import { createPortal } from 'react-dom';
+import { Editor, Range, Text, Transforms } from 'slate';
+import { ReactEditor, useSlate } from 'slate-react';
+
+export const EditorToolbar = () => {
+  const bn = 'beatmap-discussion-editor-toolbar';
+  const ref = useRef({} as HTMLDivElement);
+  const editor = useSlate();
+
+  useEffect(() => {
```

Comment:
Yeah, it's based on [this example](https://github.com/ianstormtaylor/slate/blob/master/site/examples/hovering-toolbar.js)...

I guess the difference is I've since extracted it out of the parent (Editor) because it was getting too big... but they're kinda coupled in their behaviour?